File: python/src/asqav/doors.py
"""Standards-interop "doors": one Asqav receipt, many native envelopes.

Each door wraps the SAME receipt object verbatim into a standard envelope an agent
framework already consumes: a W3C Verifiable Credential, a CloudEvents event, an OTel
GenAI attribute map, a C2PA assertion, and an ERC-8004 validationRequest shape.

The inner receipt is byte-identical across every door: the receipt dict is embedded
unchanged (VC ``credentialSubject``, CloudEvents ``data``, C2PA ``data.receipt``,
ERC-8004 ``receipt``) or, for the flat OTel attribute map, carried as its JCS string
under ``asqav.receipt``. Every door has an inverse that returns the exact same receipt.

All functions are pure and offline: no signing, no network, no mutation of the input,
no wall-clock reads. Timestamps and ids come only from the receipt itself, so the same
input always yields the same envelope in both the Python and TypeScript SDKs.

This is presentation only. A door does NOT re-sign; the authoritative Asqav signature
stays inside the embedded receipt. A generic VC / C2PA verifier reads the shape but must
use an Asqav-aware verifier to check the signature.
"""
from __future__ import annotations

import hashlib
import json
from typing import Any

from ._jcs import canonical_json

__all__ = [
    "canonical_receipt_bytes",
    "receipt_digest",
    "to_w3c_vc",
    "to_cloudevent",
    "to_otel_genai_attributes",
    "to_c2pa_assertion",
    "to_erc8004_validation_request",
    "wrap_all",
    "receipt_from_vc",
    "receipt_from_cloudevent",
    "receipt_from_otel_genai_attributes",
    "receipt_from_c2pa_assertion",
    "receipt_from_erc8004_validation_request",
    "extract_receipt",
]

VC_V2_CONTEXT = "https://www.w3.org/ns/credentials/v2"
ASQAV_DOORS_CONTEXT = "https://asqav.com/context/doors/v1"
VC_CREDENTIAL_TYPE = ["VerifiableCredential", "AsqavReceiptCredential"]
VC_PROOF_TYPE = "AsqavReceiptSignature2026"

CLOUDEVENT_TYPE = "com.asqav.receipt.v1"
CLOUDEVENT_DIGEST_ATTR = "asqavreceiptdigest"

C2PA_ASSERTION_LABEL = "com.asqav.receipt.v1"

OTEL_RECEIPT_ATTR = "asqav.receipt"
OTEL_DIGEST_ATTR = "asqav.receipt.digest"
OTEL_SIGNATURE_ATTR = "asqav.signature"

ERC8004_ZERO_ADDRESS = "0x" + "00" * 20


def canonical_receipt_bytes(receipt: dict) -> bytes:
    """Return the JCS canonical bytes of ``receipt``; the same bytes both SDKs hash."""
    return canonical_json(receipt)


def receipt_digest(receipt: dict) -> str:
    """``sha256:<hex>`` over the raw canonical receipt bytes (the C2PA raw-bytes rule)."""
    return "sha256:" + hashlib.sha256(canonical_receipt_bytes(receipt)).hexdigest()


def _require_dict(receipt: Any) -> dict:
    if not isinstance(receipt, dict):
        raise TypeError("receipt must be a parsed JSON object (dict)")
    return receipt


def _payload(receipt: dict) -> dict:
    payload = receipt.get("payload")
    return payload if isinstance(payload, dict) else receipt


def _signature_bits(receipt: dict) -> tuple[Any, Any, Any]:
    """Return ``(sig, alg, kid)`` for compliance-mode or flat hash-mode receipts."""
    sig = receipt.get("signature")
    if isinstance(sig, dict):
        return sig.get("sig"), sig.get("alg"), sig.get("kid")
    if sig is not None:
        return sig, receipt.get("algorithm"), receipt.get("key_id")
    return (
        receipt.get("signature_b64"),
        receipt.get("algorithm"),
        receipt.get("key_id"),
    )


def _field(receipt: dict, *keys: str) -> Any:
    payload = _payload(receipt)
    for key in keys:
        if payload.get(key) is not None:
            return payload[key]
        if receipt.get(key) is not None:
            return receipt[key]
    return None


def _agent_id(receipt: dict) -> Any:
    return _field(receipt, "agent_id")


def _action_id(receipt: dict) -> Any:
    return _field(receipt, "action_id", "action_ref")


def _issuer_id(receipt: dict) -> Any:
    return _field(receipt, "issuer_id", "org_id")


def _timestamp(receipt: dict) -> Any:
    return _field(receipt, "issued_at", "server_timestamp")


def _verify_urn(receipt: dict) -> str:
    """A deterministic offline pointer to the receipt by digest. Not a live endpoint."""
    return "urn:asqav:receipt:" + receipt_digest(receipt).split(":", 1)[1]


def to_w3c_vc(receipt: dict, *, issuer: str | None = None) -> dict:
    """Wrap the receipt as a W3C VC 2.0 with the whole receipt as ``credentialSubject``.

    ``proof`` is the non-standard ``AsqavReceiptSignature2026`` referencing the receipt's
    own signature. It is not a registered VC cryptosuite, so a generic VC verifier must
    decline it and defer to an Asqav-aware verifier.
    """
    _require_dict(receipt)
    sig, alg, kid = _signature_bits(receipt)
    issuer_id = issuer if issuer is not None else _issuer_id(receipt)
    vc: dict[str, Any] = {
        "@context": [VC_V2_CONTEXT, ASQAV_DOORS_CONTEXT],
        "type": list(VC_CREDENTIAL_TYPE),
        "issuer": issuer_id if issuer_id is not None else "https://asqav.com",
        "credentialSubject": receipt,
        "proof": {
            "type": VC_PROOF_TYPE,
            "proofPurpose": "assertionMethod",
            "signatureAlgorithm": alg,
            "verificationKeyId": kid,
            "signatureValue": sig,
            "receiptDigest": receipt_digest(receipt),
        },
    }
    ts = _timestamp(receipt)
    if ts is not None:
        vc["validFrom"] = ts
    return vc


def to_cloudevent(receipt: dict) -> dict:
    """Wrap the receipt as a CloudEvents 1.0 event with the receipt as ``data``.

    Adds one Asqav extension context attribute, ``asqavreceiptdigest``, binding the
    envelope to the exact receipt bytes.
    """
    _require_dict(receipt)
    agent = _agent_id(receipt)
    action = _action_id(receipt)
    event: dict[str, Any] = {
        "specversion": "1.0",
        "id": str(action) if action is not None else receipt_digest(receipt),
        "source": "urn:asqav:agent:" + str(agent) if agent is not None else "urn:asqav",
        "type": CLOUDEVENT_TYPE,
        "datacontenttype": "application/json",
        CLOUDEVENT_DIGEST_ATTR: receipt_digest(receipt),
        "data": receipt,
    }
    ts = _timestamp(receipt)
    if ts is not None:
        event["time"] = ts
    return event


def to_otel_genai_attributes(receipt: dict) -> dict:
    """Map the receipt to an OTel GenAI ``gen_ai.*`` attribute dict.

    The byte-identical inner receipt rides as the ``asqav.receipt`` JCS string so a
    consumer recovers the exact receipt from the attribute map.
    """
    _require_dict(receipt)
    sig, _alg, _kid = _signature_bits(receipt)
    attrs: dict[str, Any] = {}
    op = _field(receipt, "type", "action_type")
    if op is not None:
        attrs["gen_ai.operation.name"] = op
    agent = _agent_id(receipt)
    if agent is not None:
        attrs["gen_ai.agent.id"] = agent
    tool = _field(receipt, "tool_name")
    if tool is not None:
        attrs["gen_ai.tool.name"] = tool
    call_id = _action_id(receipt)
    if call_id is not None:
        attrs["gen_ai.tool.call.id"] = call_id
    conversation = _field(receipt, "session_id")
    if conversation is not None:
        attrs["gen_ai.conversation.id"] = conversation
    attrs[OTEL_RECEIPT_ATTR] = canonical_receipt_bytes(receipt).decode("utf-8")
    attrs[OTEL_DIGEST_ATTR] = receipt_digest(receipt)
    if sig is not None:
        attrs[OTEL_SIGNATURE_ATTR] = sig
    return attrs


def to_c2pa_assertion(receipt: dict) -> dict:
    """Wrap the receipt as a C2PA third-party assertion using the raw-bytes rule.

    ``hash_binding.hash`` is sha256 over the raw canonical receipt bytes directly, not a
    hash of a pre-computed digest, so it is the asset's own hash.
    """
    _require_dict(receipt)
    return {
        "label": C2PA_ASSERTION_LABEL,
        "data": {
            "receipt": receipt,
            "verify": _verify_urn(receipt),
        },
        "hash_binding": {
            "alg": "sha256",
            "rule": "raw-bytes",
            "hash": hashlib.sha256(canonical_receipt_bytes(receipt)).hexdigest(),
        },
    }


def to_erc8004_validation_request(
    receipt: dict,
    *,
    validator: str | None = None,
    agent_id: str | None = None,
) -> dict:
    """Wrap the receipt as an ERC-8004 ``validationRequest`` call shape.

    ``requestHash`` is a bytes32 sha256 over the canonical receipt bytes. The full receipt
    rides under ``receipt`` so the inner object stays byte-identical to the other doors.
    """
    _require_dict(receipt)
    resolved_agent = agent_id if agent_id is not None else _agent_id(receipt)
    digest_hex = hashlib.sha256(canonical_receipt_bytes(receipt)).hexdigest()
    return {
        "function": "validationRequest",
        "args": {
            "validator": validator if validator is not None else ERC8004_ZERO_ADDRESS,
            "agentId": resolved_agent,
            "requestURI": _verify_urn(receipt),
            "requestHash": "0x" + digest_hex,
        },
        "hashAlgorithm": "sha256",
        "receipt": receipt,
    }


def wrap_all(receipt: dict) -> dict:
    """Return every door for ``receipt`` keyed by standard name; one inner object, N skins."""
    _require_dict(receipt)
    return {
        "w3c_vc": to_w3c_vc(receipt),
        "cloudevent": to_cloudevent(receipt),
        "otel_genai": to_otel_genai_attributes(receipt),
        "c2pa": to_c2pa_assertion(receipt),
        "erc8004": to_erc8004_validation_request(receipt),
    }


def receipt_from_vc(vc: dict) -> dict:
    """Recover the inner receipt from a VC door."""
    return vc["credentialSubject"]


def receipt_from_cloudevent(event: dict) -> dict:
    """Recover the inner receipt from a CloudEvents door."""
    return event["data"]


def receipt_from_otel_genai_attributes(attrs: dict) -> dict:
    """Recover the inner receipt from an OTel GenAI attribute map."""
    return json.loads(attrs[OTEL_RECEIPT_ATTR])


def receipt_from_c2pa_assertion(assertion: dict) -> dict:
    """Recover the inner receipt from a C2PA door."""
    return assertion["data"]["receipt"]


def receipt_from_erc8004_validation_request(request: dict) -> dict:
    """Recover the inner receipt from an ERC-8004 door."""
    return request["receipt"]


def extract_receipt(envelope: dict) -> dict:
    """Recover the inner receipt from any door envelope, detected by shape."""
    if not isinstance(envelope, dict):
        raise TypeError("envelope must be a dict")
    if "credentialSubject" in envelope:
        return receipt_from_vc(envelope)
    if envelope.get("type") == CLOUDEVENT_TYPE or ("specversion" in envelope and "data" in envelope):
        return receipt_from_cloudevent(envelope)
    if OTEL_RECEIPT_ATTR in envelope:
        return receipt_from_otel_genai_attributes(envelope)
    if envelope.get("label") == C2PA_ASSERTION_LABEL:
        return receipt_from_c2pa_assertion(envelope)
    if envelope.get("function") == "validationRequest":
        return receipt_from_erc8004_validation_request(envelope)
    raise ValueError("unrecognised door envelope")

File: typescript/src/doors.ts
// Standards-interop doors: one Asqav receipt, N native envelopes (W3C VC, CloudEvents,
// OTel GenAI, C2PA, ERC-8004). Inner receipt is byte-identical across every door.
// Pure and offline: no signing, no network, no mutation, no clock reads. See doors.py.
import { createHash } from "node:crypto";

import { canonicalJson } from "./jcs.js";

export type Receipt = Record<string, unknown>;

export const VC_V2_CONTEXT = "https://www.w3.org/ns/credentials/v2";
export const ASQAV_DOORS_CONTEXT = "https://asqav.com/context/doors/v1";
export const VC_CREDENTIAL_TYPE = ["VerifiableCredential", "AsqavReceiptCredential"];
export const VC_PROOF_TYPE = "AsqavReceiptSignature2026";

export const CLOUDEVENT_TYPE = "com.asqav.receipt.v1";
export const CLOUDEVENT_DIGEST_ATTR = "asqavreceiptdigest";

export const C2PA_ASSERTION_LABEL = "com.asqav.receipt.v1";

export const OTEL_RECEIPT_ATTR = "asqav.receipt";
export const OTEL_DIGEST_ATTR = "asqav.receipt.digest";
export const OTEL_SIGNATURE_ATTR = "asqav.signature";

export const ERC8004_ZERO_ADDRESS = "0x" + "00".repeat(20);

/** JCS canonical string of the receipt; the same bytes both SDKs hash. */
function canonicalString(receipt: Receipt): string {
  return new TextDecoder().decode(canonicalJson(receipt));
}

/** JCS canonical bytes of the receipt. */
export function canonicalReceiptBytes(receipt: Receipt): Uint8Array {
  return canonicalJson(receipt);
}

/** `sha256:<hex>` over the raw canonical receipt bytes (the C2PA raw-bytes rule). */
export function receiptDigest(receipt: Receipt): string {
  return "sha256:" + sha256Hex(receipt);
}

function sha256Hex(receipt: Receipt): string {
  return createHash("sha256").update(Buffer.from(canonicalReceiptBytes(receipt))).digest("hex");
}

function requireObject(receipt: unknown): Receipt {
  if (receipt === null || typeof receipt !== "object" || Array.isArray(receipt)) {
    throw new TypeError("receipt must be a parsed JSON object");
  }
  return receipt as Receipt;
}

function payload(receipt: Receipt): Receipt {
  const p = receipt.payload;
  return p !== null && typeof p === "object" && !Array.isArray(p) ? (p as Receipt) : receipt;
}

function signatureBits(receipt: Receipt): { sig: unknown; alg: unknown; kid: unknown } {
  const sig = receipt.signature;
  if (sig !== null && typeof sig === "object" && !Array.isArray(sig)) {
    const s = sig as Receipt;
    return { sig: s.sig, alg: s.alg, kid: s.kid };
  }
  if (sig !== undefined && sig !== null) {
    return { sig, alg: receipt.algorithm, kid: receipt.key_id };
  }
  return { sig: receipt.signature_b64, alg: receipt.algorithm, kid: receipt.key_id };
}

function field(receipt: Receipt, ...keys: string[]): unknown {
  const p = payload(receipt);
  for (const key of keys) {
    if (p[key] !== undefined && p[key] !== null) return p[key];
    if (receipt[key] !== undefined && receipt[key] !== null) return receipt[key];
  }
  return undefined;
}

const agentId = (r: Receipt) => field(r, "agent_id");
const actionId = (r: Receipt) => field(r, "action_id", "action_ref");
const issuerId = (r: Receipt) => field(r, "issuer_id", "org_id");
const timestamp = (r: Receipt) => field(r, "issued_at", "server_timestamp");

/** Deterministic offline pointer to the receipt by digest. Not a live endpoint. */
function verifyUrn(receipt: Receipt): string {
  return "urn:asqav:receipt:" + sha256Hex(receipt);
}

/** Wrap the receipt as a W3C VC 2.0 with the whole receipt as `credentialSubject`. */
export function toW3cVc(receipt: Receipt, opts: { issuer?: string } = {}): Receipt {
  requireObject(receipt);
  const { sig, alg, kid } = signatureBits(receipt);
  const issuer = opts.issuer !== undefined ? opts.issuer : issuerId(receipt);
  const vc: Receipt = {
    "@context": [VC_V2_CONTEXT, ASQAV_DOORS_CONTEXT],
    type: [...VC_CREDENTIAL_TYPE],
    issuer: issuer !== undefined && issuer !== null ? issuer : "https://asqav.com",
    credentialSubject: receipt,
    proof: {
      type: VC_PROOF_TYPE,
      proofPurpose: "assertionMethod",
      signatureAlgorithm: alg ?? null,
      verificationKeyId: kid ?? null,
      signatureValue: sig ?? null,
      receiptDigest: receiptDigest(receipt),
    },
  };
  const ts = timestamp(receipt);
  if (ts !== undefined && ts !== null) vc.validFrom = ts;
  return vc;
}

/** Wrap the receipt as a CloudEvents 1.0 event with the receipt as `data`. */
export function toCloudEvent(receipt: Receipt): Receipt {
  requireObject(receipt);
  const agent = agentId(receipt);
  const action = actionId(receipt);
  const event: Receipt = {
    specversion: "1.0",
    id: action !== undefined && action !== null ? String(action) : receiptDigest(receipt),
    source:
      agent !== undefined && agent !== null ? "urn:asqav:agent:" + String(agent) : "urn:asqav",
    type: CLOUDEVENT_TYPE,
    datacontenttype: "application/json",
    [CLOUDEVENT_DIGEST_ATTR]: receiptDigest(receipt),
    data: receipt,
  };
  const ts = timestamp(receipt);
  if (ts !== undefined && ts !== null) event.time = ts;
  return event;
}

/** Map the receipt to an OTel GenAI `gen_ai.*` attribute object. */
export function toOtelGenaiAttributes(receipt: Receipt): Receipt {
  requireObject(receipt);
  const { sig } = signatureBits(receipt);
  const attrs: Receipt = {};
  const op = field(receipt, "type", "action_type");
  if (op !== undefined && op !== null) attrs["gen_ai.operation.name"] = op;
  const agent = agentId(receipt);
  if (agent !== undefined && agent !== null) attrs["gen_ai.agent.id"] = agent;
  const tool = field(receipt, "tool_name");
  if (tool !== undefined && tool !== null) attrs["gen_ai.tool.name"] = tool;
  const callId = actionId(receipt);
  if (callId !== undefined && callId !== null) attrs["gen_ai.tool.call.id"] = callId;
  const conversation = field(receipt, "session_id");
  if (conversation !== undefined && conversation !== null) {
    attrs["gen_ai.conversation.id"] = conversation;
  }
  attrs[OTEL_RECEIPT_ATTR] = canonicalString(receipt);
  attrs[OTEL_DIGEST_ATTR] = receiptDigest(receipt);
  if (sig !== undefined && sig !== null) attrs[OTEL_SIGNATURE_ATTR] = sig;
  return attrs;
}

/** Wrap the receipt as a C2PA third-party assertion using the raw-bytes rule. */
export function toC2paAssertion(receipt: Receipt): Receipt {
  requireObject(receipt);
  return {
    label: C2PA_ASSERTION_LABEL,
    data: {
      receipt,
      verify: verifyUrn(receipt),
    },
    hash_binding: {
      alg: "sha256",
      rule: "raw-bytes",
      hash: sha256Hex(receipt),
    },
  };
}

/**
 * asqav sha256 `receiptCommitment`, not the ERC-8004 keccak256 `requestHash`. See doors.py.
 */
export function toErc8004ValidationRequest(
  receipt: Receipt,
  opts: { validator?: string; agentId?: string } = {},
): Receipt {
  requireObject(receipt);
  const resolvedAgent = opts.agentId !== undefined ? opts.agentId : agentId(receipt);
  return {
    function: "validationRequest",
    args: {
      validator: opts.validator !== undefined ? opts.validator : ERC8004_ZERO_ADDRESS,
      agentId: resolvedAgent ?? null,
      requestURI: verifyUrn(receipt),
      receiptCommitment: "0x" + sha256Hex(receipt),
    },
    hashAlgorithm: "asqav-commitment-sha256",
    receipt,
  };
}

/** Return every door keyed by standard name; one inner object, N skins. */
export function wrapAll(receipt: Receipt): Receipt {
  requireObject(receipt);
  return {
    w3c_vc: toW3cVc(receipt),
    cloudevent: toCloudEvent(receipt),
    otel_genai: toOtelGenaiAttributes(receipt),
    c2pa: toC2paAssertion(receipt),
    erc8004: toErc8004ValidationRequest(receipt),
  };
}

export function receiptFromVc(vc: Receipt): Receipt {
  return vc.credentialSubject as Receipt;
}

export function receiptFromCloudEvent(event: Receipt): Receipt {
  return event.data as Receipt;
}

export function receiptFromOtelGenaiAttributes(attrs: Receipt): Receipt {
  return JSON.parse(attrs[OTEL_RECEIPT_ATTR] as string) as Receipt;
}

export function receiptFromC2paAssertion(assertion: Receipt): Receipt {
  return (assertion.data as Receipt).receipt as Receipt;
}

export function receiptFromErc8004ValidationRequest(request: Receipt): Receipt {
  return request.receipt as Receipt;
}

/** Recover the inner receipt from any door envelope, detected by shape. */
export function extractReceipt(envelope: Receipt): Receipt {
  requireObject(envelope);
  if ("credentialSubject" in envelope) return receiptFromVc(envelope);
  if (envelope.type === CLOUDEVENT_TYPE || ("specversion" in envelope && "data" in envelope)) {
    return receiptFromCloudEvent(envelope);
  }
  if (OTEL_RECEIPT_ATTR in envelope) return receiptFromOtelGenaiAttributes(envelope);
  if (envelope.label === C2PA_ASSERTION_LABEL) return receiptFromC2paAssertion(envelope);
  if (envelope.function === "validationRequest") {
    return receiptFromErc8004ValidationRequest(envelope);
  }
  throw new Error("unrecognised door envelope");
}
